### PR TITLE
WIP/RFC:  Add tests for ancestral.py

### PR DESCRIPTION
**Pull request by vanguard737**
Per #476, this PR adds some initial tests for ancestral.py.

tl;dr It ain't pretty, and these aren't unit tests - more like a stripped-down integration test using the data from `tests/builds/runner.sh`.  But it passes, both locally and on Travis, and gets `ancestral.py`'s code coverage from 8% to 67%.  Possibly (definitely?) not suitable for merging as-is.  But, with *a* test in place, the door's open to write some better tests and perhaps refactor a bit.

I found it kind of hard to write unit tests for `ancestral.py` in its current state.  (In particular, ancestral.run() does a lot of stuff - it constructs some objects internally that get written to disk before I can assert anything about them.)  But, I wasn't comfortable making any changes without some kind of regression test.  And, I wanted something a little more tightly-controlled than just running `tests/builds/runner.sh` in its entirety.  So, I did the following:

- Executed `tests/builds/runner.sh` with `snakefile --printshellcmds` turned on.  This pointed me to the builds that include calls to `augur ancestral`, which are:
	- add_to_alignment
	- tb
	- tb_drm
	- zika

- For each of those builds:
	- Identified the input files to, and output file(s) from, the `augur ancestral` call.  Got these under source control in `tests/data/ancestral/<build name>/in` and `tests/data/ancestral/<build name>/expected`, respectively.
	- Reproduced the params for the `augur ancestral` call as an argparse.Namespace() object.
	- Passed that into ancestral.run().
	- Compared the resulting file(s) to the expected results in tests/data/ancestral/<build name>/expected.  (Typically nt_muts.json; 'tb' and 'tb_drm' also produce a VCF file.)  Files match, tests pass.

...And that's it, for the moment.  I've left `ancestral.py` untouched for the purposes of this PR.  Reasonable next steps might include:

- Actual unit tests for `ancestral_sequence_inference()` and `collect_mutations_and_sequences()`.  In retrospect, I should probably have done this first.  But that requires a suitable fixture for Bio.Phylo.Tree and Bio.Align.MultipleSeqAlignment - and I wasn't sure where to get those, except by digging 'em out of the snakemake builds.  Maybe there's a better option?
- Refactoring `run()` - Might be a good idea to create the output_node_data outside of run(), and check that it makes sense before writing to disk.  
- Other scenarios - The snakemake builds don't include any cases where:
	- `inference = "marginal"`, or
	- `infer_ambiguous = False`, or
	- `keep_overhangs = True`
...so, currently, these tests don't either.

Any comments or feedback, please fire away.  As an Internet rando, I'm sure I'm missing lots of context here, so all feedback is warmly appreciated!

### CODE REVIEW

**Comment by jameshadfield** [comment]
Really exiciting to see some tests covering augur sub-commands and provide a path to add these for more sub-commands 🎉 will try to check back here periodically

**Comment by tsibley** [comment]
Thanks for this work, @vanguard737! I appreciate you diving into these tests.

I think this sort of CLI testing would be a lot clearer and easier with [cram](https://bitheap.org/cram/) (probably plugged into pytest using [pytest-cram](https://pypi.org/project/pytest-cram/)). I've raised this alternate solution in another issue/PR or two as well as internally amongst the team, but I don't think there's been much time to actually consider it yet. Now might be a good time though, since we're considering this PR. @huddlej @jameshadfield tagging you here for your :eyes:.

Here's what one test case in this PR would look like in cram, for example:

```
Reproduce the behavior of the 'augur ancestral' CLI command, as implemented in
the 'add_to_alignment' test build:

  $ TEST_DATA_DIR="$TESTDIR/data/ancestral/add_to_alignment/"
  $ mkdir -p "$TMP/out"

  $ augur ancestral \
  >   --tree "$TEST_DATA_DIR/in/tree.nwk" \
  >   --alignment "$TEST_DATA_DIR/in/aligned.fasta" \
  >   --output-node-data "$TMP/out/nt_muts.json" \
  >   --inference joint \
  >   >/dev/null

  $ diff -u "$TEST_DATA_DIR/expected/nt_muts.json" "$TMP/out/nt_muts.json"
```

**Comment by vanguard737** [comment]
@tsibley @jameshadfield Thanks for the feedback!  Sorry for the delayed response here, I seem to have gone down a bit of a rabbit-hole with [cram](https://bitheap.org/cram/) and [pytest-cram](https://pypi.org/project/pytest-cram/).  I'm putting together a separate PR for that - it'll probably be ready for initial submission tonight or tomorrow.  Will plan to tag you guys (and @huddlej) on that when it's all set.

**Comment by vanguard737** [comment]
Mmph, my optimism has gotten the better of me.  PR not quite ready yet, but work is in progress [here](https://github.com/alpha29/augur/tree/test_cli_cram).  Ongoing...

**Comment by genehack** [comment]
I am closing this pull request. It has been untouched for over four years, and has merge conflicts with the current codebase, which did end up moving in the direction of `cram` for integration tests.  

We're trying to get a good handle on our open issues and PRs, so that they can serve as a reliable backlog of things we actually will work on. If you still think this is a worthwhile contribution, please reopen (and update, if necessary) the PR and explain why, thank.
